transformer_engine/pytorch/float8_tensor.py

```diff
@@ -0,0 +1,617 @@
+        *,
+        data: torch.Tensor,
+        **kwargs,
+    ) -> Float8Tensor:
+        """Use attributes of a Float8Tensor to create another Float8Tensor
+
+        See constructor for list of keyword arguments.
+
+        """
+        default_kwargs = dict(
+            fp8_meta=tensor._fp8_meta,
+            fp8_meta_forward=tensor._fp8_meta_forward,
+            fp8_meta_index=tensor._fp8_meta_index,
+            fp8_dtype=tensor._fp8_dtype,
+            fp8_scale_inv=tensor._scale_inv,
+            dtype=tensor.dtype,
+        )
+        for key, val in default_kwargs.items():
+            if key not in kwargs:
+                kwargs[key] = val
+        return Float8Tensor(data=data, **kwargs)
+
+    def __repr__(self):
+        return (
+            "Float8Tensor("
+            f"fp8_dtype={self._fp8_dtype}, "
+            f"scale_inv={self._scale_inv.item()}, "
+            f"data={self.from_float8(dtype=self.dtype)}"
+            ")"
+        )
+
+    def from_float8(self, dtype: Optional[torch.dtype] = None) -> torch.Tensor:
+        """Construct plain PyTorch tensor from Float8Tensor"""
+        return _FromFloat8Func.apply(self, dtype)
+
+    @classmethod
+    def to_float8(
+        cls,
+        tensor: torch.Tensor,
+        *,
+        fp8_meta: Optional[Dict[str, Any]] = None,
+        fp8_meta_forward: bool = True,
+        fp8_meta_index: Optional[int] = None,
+        fp8_dtype: Optional[tex.DType] = None,
+        scale: Optional[torch.Tensor] = None,
+        amax: Optional[torch.Tensor] = None,
+        scale_inv: Optional[torch.Tensor] = None,
+    ):
+        """Construct Float8Tensor from plain PyTorch tensor"""
+        return _ToFloat8Func.apply(
+            tensor,
+            fp8_meta,
+            fp8_meta_forward,
+            fp8_meta_index,
+            fp8_dtype,
+            scale,
+            amax,
+            scale_inv,
+        )
+
+    def float(self) -> torch.Tensor:
+        return self.from_float8(dtype=torch.float32)
+
+    def bfloat16(self) -> torch.Tensor:
+        return self.from_float8(dtype=torch.bfloat16)
+
+    def half(self) -> torch.Tensor:
+        return self.from_float8(dtype=torch.float16)
+
+    def cpu(self) -> torch.Tensor:
+        return self.from_float8().cpu()
+
+    def clone(self) -> Float8Tensor:
+        return _IdentityFunc.apply(self, {"data": self._data.detach().clone()})
+
+    def expand_as(self, other: torch.Tensor):
+        if other is self:
+            # Note: expand_as is hackily used to create dummy autograd nodes
+            # and access the backward graph (see
+            # https://github.com/pytorch/pytorch/blob/238fb660851268f44ff88127887041fea352fe48/torch/nn/parallel/distributed.py#L1026).
+            # We equally hackily add a dummy function to handle this
+            # case.
+            return _IdentityFunc.apply(self)
+        return super().expand_as(other)
+
+    def transpose(self, dim0: int = 0, dim1: int = 1) -> Float8Tensor:
```

Comment:
Maybe add an optional argument here that says whether to cache the result?

Reply:
Alternatively, we could do something like:
```python
def transpose(self):
    if self._transpose is not None:
        return self._transpose
    return super().transpose()
```
The cached transpose should be allocated and initialized externally, e.g. in the TE modules.